gdcdatamodel/validators/link_validator_parser.py

```diff
@@ -0,0 +1,185 @@
+        self.update_exclusive_mask(child)
+
+    def add_child_link(self, child):
+        self.update_parent(child)
+
+
+class LinkWithExclusiveMask(object):
+    def __init__(self, item):
+        self.name = item.name
+        self.exclusive_mask = item.exclusive_mask
+        self.exclusive_links = item.exclusive_links
+        self.multiplicity = item.schema.get('multiplicity')
+        self.backref = item.schema.get('backref')
+
+
+class ExistingMasks(object):
+    def __init__(self, code, existing_mask, existing_links):
+        self.code = code
+        self.existing_mask = existing_mask
+        self.existing_links = existing_links
+
+
+class RequiredMask(object):
+    def __init__(self):
+        self.list_required_links = []
+        self.group_required = []
+        self.required_mask = 0
+
+
+def create_subgroup_validators(node_label):
+    link_items = []
+    leaves = []
+    for link in gdcdictionary.schema[node_label]['links']:
+        if 'name' in link:
+            l_item = SubgroupParsingNode(None, link, 1 << len(leaves))
+            leaves.append(l_item)
+        else:
+            l_item = SubgroupParsingNode(None, link)
+        link_items.append(l_item)
+
+    return build_tree_of_link_items(link_items, leaves)
+
+
+def build_tree_of_link_items(link_items, leaves):
+    final_required = []
+    current_pos = 0
+    while current_pos < len(link_items):
+        l_item = link_items[current_pos]
+        children_required_count = 0
+        if 'subgroup' in l_item.schema:
+            for link in l_item.schema['subgroup']:
+                if 'subgroup' in link:
+                    new_l_item = SubgroupParsingNode(l_item, link)
+                else:
+                    new_l_item = SubgroupParsingNode(l_item, link, 1 << len(leaves))
+                    l_item.add_child_link(new_l_item)
+                    leaves.append(new_l_item)
+
+                link_items.append(new_l_item)
+
+                if l_item.required and new_l_item.required:
+                    children_required_count += 1
+        if l_item.all_ancestors_required and l_item.required and children_required_count == 0:
+            final_required.append(l_item)
+        current_pos += 1
+
+    required_validator, base_mask = create_required_mask(final_required)
+    existing_list = create_existing_list(link_items, base_mask)
+    exclusive_list = create_exclusive_list(leaves)
+
+    return required_validator, existing_list, exclusive_list
+
+
+def create_exclusive_list(leaves):
+    if len(leaves) == 0:
+        return []
+
+    current = leaves[len(leaves) - 1]
+    leaves_checks = deque([LinkWithExclusiveMask(current)])
+    i = len(leaves) - 2
+    while i >= 0:
+        leaf = leaves[i]
+        if leaf.parent == current.parent:
+            leaf.exclusive_links = current.exclusive_links
+        else:
+            current = leaf
+        leaves_checks.appendleft(LinkWithExclusiveMask(leaf))
+        i -= 1
+    return list(leaves_checks)
```

Comment:
I'm having some difficulty following where the first `exclusive_links` comes from, it feels like it would be better to just initialize them somewhere else, during the `__init__` maybe?